### PR TITLE
Add clippy check to travis 

### PR DESCRIPTION
**Pull request by chrisduerr**
This commit adds clippy as a required step of the build process. To make
this possible, all existing clippy issues have been resolved.

I think making clippy part of the travis CI run would be a great step,
but if this is not desired, I can also take that part out of this and
add just the fixes for all current clippy issues.

### CODE REVIEW

File: src/logging.rs [hunk 1]
```diff
@@ -28,6 +28,7 @@ pub struct Logger<T> {
 }
 
 impl<T: Send + io::Write> Logger<T> {
+    #[cfg_attr(feature = "clippy", allow(new_ret_no_self))]
```
Comment (jwilm) [inline src/logging.rs]: Let's add a comment here pointing to the issue tracking this broken lint: https://github.com/rust-lang-nursery/rust-clippy/issues/734

File: src/renderer/mod.rs [hunk 2]
```diff
@@ -1125,7 +1125,7 @@ impl ShaderProgram {
             gl::GetShaderiv(shader, gl::COMPILE_STATUS, &mut success);
         }
 
-        if success == (gl::TRUE as GLint) {
+        if success == i32::from(gl::TRUE) {
```
Comment (jwilm) [inline src/renderer/mod.rs]: This should probably be `GLint::from` rather than `i32::from`.

File: src/selection.rs [hunk 3]
```diff
@@ -136,25 +136,22 @@ impl Selection {
             Selection::Simple { ref mut region } => {
                 region.end = Anchor::new(location, side);
             },
-            Selection::Semantic { ref mut region, .. } => {
+            Selection::Semantic { ref mut region, .. } | Selection::Lines { ref mut region, .. } => {
```
Comment (jwilm) [inline src/selection.rs]: Break line after pipe

File: .travis.yml [hunk 4]
```diff
@@ -11,10 +11,17 @@ rust:
...
-  allow_failures:
+  exclude:
     - rust: nightly
+      env: VERSION="stable"
+    - rust: stable
+      env: VERSION="nightly"
 
 script:
-  - cargo test --no-default-features
+  - ./build.sh
```
Comment (jwilm) [inline .travis.yml]: Let's try and keep all the shell commands in the `script` section rather than adding `build.sh`. The build.sh script can be

```yml
- if [[ $VERSION == "nightly" ]]; then cargo test --no-default-features --features "clippy"; fi
- if [[ $VERSION != "nightly" ]]; then cargo test --no-default-features; fi
```

File: src/ansi.rs [hunk 5]
```diff
@@ -360,6 +360,9 @@ pub enum CursorStyle {
 
     /// Cursor is a vertical bar `⎸`
     Beam,
+
+    /// Cursor is a box like `☐`
+    Box,
```
Comment (jwilm) [inline src/ansi.rs]: This looks new in this PR. Can you pull it out into a separate patch? Small nit also -- it would be nice to name this `Hollow` or `Rect` or something other than `Box`.

File: src/term/mod.rs [hunk 6]
```diff
@@ -988,6 +990,11 @@ impl Term {
     ) -> RenderableCellsIter {
         let selection = selection.and_then(|s| s.to_span(self))
             .map(|span| span.to_range());
+        let cursor = if window_focused {
```
Comment (jwilm) [inline src/term/mod.rs]: Unrelated to this PR

File: src/tty.rs [hunk 7]
```diff
@@ -70,8 +70,8 @@ fn openpty(rows: u8, cols: u8) -> (c_int, c_int) {
     let mut slave: c_int = 0;
 
     let win = winsize {
-        ws_row: rows as libc::c_ushort,
-        ws_col: cols as libc::c_ushort,
+        ws_row: u16::from(rows),
```
Comment (jwilm) [inline src/tty.rs]: `c_ushort::from`

File: src/tty.rs [hunk 8]
```diff
@@ -93,8 +93,8 @@ fn openpty(rows: u8, cols: u8) -> (c_int, c_int) {
     let mut slave: c_int = 0;
 
     let mut win = winsize {
-        ws_row: rows as libc::c_ushort,
-        ws_col: cols as libc::c_ushort,
+        ws_row: u16::from(rows),
```
Comment (jwilm) [inline src/tty.rs]: `c_ushort::from`

File: src/tty.rs [hunk 9]
```diff
@@ -113,6 +113,7 @@ fn openpty(rows: u8, cols: u8) -> (c_int, c_int) {
 /// Really only needed on BSD, but should be fine elsewhere
 fn set_controlling_terminal(fd: c_int) {
     let res = unsafe {
+        #[allow(clippy)]
```
Comment (jwilm) [inline src/tty.rs]: What didn't clippy like about this? It would be nice to just disable a single lint rather than all of clippy for this line.

Comment (chrisduerr) [reply src/tty.rs]: This was a problem with multiple platforms. Because `TIOSCTTY` has a different type on macos and linux. So the conversion fails on clippy because it's either an unnecessary conversion, or an incorrect type.

Comment (chrisduerr) [reply src/tty.rs]: I've tried to use `if cfg!(target_os = "macos")` but I wasn't able to make it work. Clippy still complained about the branch that shouldn't be reached on linux.

File: src/term/mod.rs (hunk 6 above)
Comment (chrisduerr) [reply src/term/mod.rs]: These changes were made to reduce the amount of parameters of the method below clippy's threshold.

**Comment by chrisduerr** [comment]
Rebased this on top of master, should be good when travis doesn't complain?

File: src/ansi.rs (hunk 5 above)
Comment (jwilm) [reply src/ansi.rs]: Or maybe `HollowBlock` would be best

File: src/tty.rs [hunk 10]
```diff
@@ -113,7 +113,10 @@ fn openpty(rows: u8, cols: u8) -> (c_int, c_int) {
 /// Really only needed on BSD, but should be fine elsewhere
 fn set_controlling_terminal(fd: c_int) {
     let res = unsafe {
-        libc::ioctl(fd, TIOCSCTTY as _, 0)
+        // Cross platform issue because on linux this is u64 as u64 (does nothing)
+        // But on macos this is u32 as u64, asking for u64::from(u32)
```
Comment (jwilm) [inline src/tty.rs]: Thanks for the comment!
